### PR TITLE
Update data load pattern to be extensible

### LINKED ISSUE #20: Create ability for Seismogram to start from_dataframe()
## Problem Summary
Currently the notebook requires the data to be in parquet format.  This is a good first step as it has implicit typing.  A logical next step is to accept the dataframe to already be loaded.

## Impact
Users may not have parquet files, and writing them just so they can be loaded is awkward (though has benefits on 'Restart and run all' situations).

## Possible Solution
Create a new class method from_dataframe that will expect the config file for data usage but will be passed the dataframe instead of loading from files.  

This will have implications on the default constructor as well; initial thoughts to be consistent would be moved to a 'from_parquet' method

### PR DESCRIPTION
**Pull request by diehlbw**
# Overview

Closes #20 

## Description of changes
Pulled the loading logic for prediction and event data out of the Seismogram class

- [ ] Should maintain functionality for loading from parquet
- [ ] Existing example notebook should work unchanged (ie Verify run_startup works as is)
- [ ] Verify that loading is also possible from preloaded dataframes.  If using the example notebook make sure that the files don't _also_ exist and update run_startup cell to pass in the loaded dataframes.

Note: Updates to allow loading from urls will be done in a follow-up PR

## Author Checklist
- [x] Linting passes; run early with [pre-commit hook](https://pre-commit.com/#install).
- [x] Tests added for new code and issue being fixed.
- [ ] Added type annotations and full numpy-style docstrings for new methods.
- [ ] Draft your news fragment in new `changelog/ISSUE.TYPE.rst` files; see [changelog/README.md](https://github.com/epic-open-source/seismometer/blob/main/changelog/README.md).

### CODE REVIEW

File: src/seismometer/data/loader/pipeline.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        """
+        Initialize a data loading pipeline of functions returning a dataframe for a Seismogram session.
+
+        Parameters
+        ----------
+        config : ConfigProvider
+            The loaded configuration object.
+        prediction_fn : ConfigOnlyHook
+            A callable taking a ConfigProvider and returning a dataframe.
+            Used to load a (predictions) dataframe based on configuration.
```
Comment (gbowlin) [inline src/seismometer/data/loader/pipeline.py:78]: Can we add comment about this being ignored if load_data is called with a dataframe?

File: src/seismometer/data/loader/pipeline.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        logger.info(f"Importing files from {self.config.config_dir}")
+
+        dataframe = self._load_predictions(prediction_obj)
+        dataframe = self.post_predict_fn(self.config, dataframe)
+        dataframe = self._add_events(dataframe, event_obj)
+
+        return dataframe
+
+    def _load_predictions(self, prediction_obj: pd.DataFrame = None):
+        """Load predictions from config or memory."""
```
Comment (gbowlin) [inline src/seismometer/data/loader/pipeline.py:133]: Can we add a comment on preference for prediction_object over prediction_fn?,

File: src/seismometer/data/loader/pipeline.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+    def _load_events(self, event_obj: pd.DataFrame = None) -> pd.DataFrame:
+        """Load events from config or memory."""
+        if (event_obj is None) and (self.event_fn is None):
+            return pd.DataFrame()
+        if event_obj is None:
+            return self.event_fn(self.config)
+        return self.event_from_memory(self.config, event_obj)
+
+
+__all__ = ["SeismogramLoader", "ConfigOnlyHook", "ConfigFrameHook", "MergeHook"]
```
Comment (gbowlin) [inline src/seismometer/data/loader/pipeline.py:158]: Should this use the @export decorator?

File: src/seismometer/data/loader/prediction.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+        if config.target in present_columns:
+            desired_columns.add(config.target)
+
+        actual_columns = desired_columns & present_columns
+        _log_column_mismatch(actual_columns, desired_columns, present_columns)
+
+        dataframe = pd.read_parquet(config.prediction_path, columns=actual_columns)
+    else:
+        dataframe = pd.read_parquet(config.prediction_path)
```
Comment (gbowlin) [inline src/seismometer/data/loader/prediction.py:45]: Since the smaller section is the else, can we make that go first?

It also matches your comment better (# no features == all features)

File: src/seismometer/data/loader/prediction.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        return
+    logger.warning(
+        "Not all requested columns are present. " + f"Missing columns are {', '.join(desired_columns-present_columns)}"
+    )
+    logger.debug(f"Requested columns are {', '.join(desired_columns)}")
+    logger.debug(f"Columns present are {', '.join(present_columns)}")
+
+
+def _rename_targets(config: ConfigProvider, dataframe: pd.DataFrame) -> pd.DataFrame:
+    """Renames the target column if alread in dataframe, to match what a event merge would produce."""
```
Comment (gbowlin) [inline src/seismometer/data/loader/prediction.py:64]:
```suggestion
    """Renames the target column if already in the dataframe, to match what a event merge would produce."""
```

File: src/seismometer/data/loader/prediction.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    # datetime precisions don't play nicely - fix to pands default
+    pred_times = dataframe.select_dtypes(include="datetime").columns
+    dataframe[pred_times] = dataframe[pred_times].astype({col: "<M8[ns]" for col in pred_times})
+
+    # Expand this to robust score prep
+    for score in config.output_list:
+        if score not in dataframe:
+            continue
+        if 50 < dataframe[score].max() <= 100:  # Assume out of 100, readjust
+            dataframe[score] /= 100
```
Comment (gbowlin) [inline src/seismometer/data/loader/prediction.py:102]: Not this PR, but this was copied from seismogram.py and it feels brittle, since many of our scores are expected to trigger at low values due to the low prevalence.

File: src/seismometer/seismogram.py [hunk 7]
```diff
@@ -72,12 +71,21 @@ def __init__(self, config_path: Optional[str | Path] = None, output_path: Option
             config_path = Path(config_path)
 
         self.cohort_cols: list[str] = []
```
Comment (gbowlin) [inline src/seismometer/seismogram.py:73]: Should this be redirected to config as well like you did for target_cols?

File: src/seismometer/seismogram.py [hunk 8]
```diff
@@ -280,9 +285,7 @@ def _load_from_config(self, config_path: str | Path):
         self.predict_time = self.config.predict_time
         self.output_list = self.config.output_list
         self.target_event = self.config.target
-
-        self._load_metadata()
-        self._load_data()
+        self._cohorts = self.config.cohorts
```
Comment (gbowlin) [inline src/seismometer/seismogram.py:288]: Rather than setting _cohorts, should we just redirect to self.config as a property?

**Review by gbowlin — COMMENTED** [review]
Done with the code side, haven't looked at tests yet.

File: docs/reference/seismogram.rst [hunk 9]
```diff
@@ -18,31 +18,33 @@ Properties
...
-    Seismogram.entity_count
-    Seismogram.feature_count
-    Seismogram.start_time
-    Seismogram.end_time
-    Seismogram.event_count
-    Seismogram.event_types_count
-    Seismogram.cohort_attribute_count
-    Seismogram.score_bins
-    Seismogram.data
+   Seismogram.events
```
Comment (gbowlin) [inline docs/reference/seismogram.rst:21]: Why the change by one space?

File: src/seismometer/_api.py [hunk 10]
```diff
@@ -626,16 +626,6 @@ def _get_info_dict(plot_help: bool) -> dict[str, str | list[str]]:
         "plot_help": plot_help,
     }
 
-    if sg.event_count > 0:
```
[the remark this answers is not in the document]
Comment (diehlbw) [reply src/seismometer/_api.py:859]: With the change in loading, the events table is not ever presented to Seismogram as a it's own object.  
So... I went the slightly lazy route and did not figure out a way of getting these statistics back into known state.

The best I think we could do is conditional counts (positive value, has-time, etc) but each of those fail to generalize completely (non-boolean | pertinent zero, value-only | never has time, etc)

File: src/seismometer/configuration/config.py [hunk 11]
```diff
@@ -54,6 +54,7 @@ def __init__(
         self._events: EventDictionary = None
         self._output_dir: Path = None
         self._output_notebook: str = ""
+        self.target_cols: list = []
```
[the remark this answers is not in the document]
Comment (diehlbw) [reply src/seismometer/configuration/config.py]: Moved this to be a comprehension on config like the other usage types

File: src/seismometer/data/loader/pipeline.py [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+ConfigFrameHook.__doc__ = """
+TypeAlias for a callable taking a ConfigProvider and a DataFrame, which returns a DataFrame.
+
+An example is ensuring types of a loaded predictions dataframe, where ConfigProvider
+provides some metadata that can then be applied to the dataframe returing the transformed object.
+"""
+
+
+MergeHook: TypeAlias = Callable[[ConfigProvider, pd.DataFrame, pd.DataFrame], pd.DataFrame]
+MergeHook.__doc__ = """
```
[the remark this answers is not in the document]
Comment (diehlbw) [reply src/seismometer/data/loader/pipeline.py:28]: Good suggestion, explicit is better

File: src/seismometer/data/loader/pipeline.py [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+
+def _passthru_framehook(config: ConfigProvider, dataframe: pd.DataFrame) -> pd.DataFrame:
+    """A generic pass-through used for hooks that don't need to modify the dataframe."""
+    return dataframe
+
+
+def _passthru_mergehook(config: ConfigProvider, eventframe: pd.DataFrame, dataframe: pd.DataFrame) -> pd.DataFrame:
+    """A generic pass-through used for hooks that don't need to modify the dataframe."""
+    return dataframe
```
[the remark this answers is not in the document]
Comment (diehlbw) [reply src/seismometer/data/loader/pipeline.py:44]: Added, but I did this in _add_events as it will skip calling the merge function in that case

File: src/seismometer/data/loader/pipeline.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+        dataframe = self._load_predictions(prediction_obj)
+        dataframe = self.post_predict_fn(self.config, dataframe)
+        dataframe = self._add_events(dataframe, event_obj)
+
+        return dataframe
+
+    def _load_predictions(self, prediction_obj: pd.DataFrame = None):
+        """Load predictions from config or memory."""
+        if (prediction_obj is None) and (self.prediction_fn is None):
+            return pd.DataFrame()
```
[the remark this answers is not in the document]
Comment (diehlbw) [reply src/seismometer/data/loader/pipeline.py]: Good call; added as a runtime error.  I'd expect this to primarily be the routers fault (no prediction loader) but when intended it will direct to specifying a frame

File: src/seismometer/data/loader/prediction.py (hunk 4 above)
Comment (diehlbw) [reply src/seismometer/data/loader/prediction.py:45]: I like it, makes me feel a tiny bit better about having an else-statement too

File: src/seismometer/data/loader/prediction.py (hunk 5 above)
Comment (diehlbw) [reply src/seismometer/data/loader/prediction.py:64]: Added to my fix commit

File: src/seismometer/data/loader/prediction.py (hunk 6 above)
Comment (diehlbw) [reply src/seismometer/data/loader/prediction.py:102]: I halved the minimum as a bandaid "improvement".  
I think this will need to be reconsidered once we have more robust support for other ranges of predictions (my first thought is we'll need a flag or enum in configuration)

File: src/seismometer/seismogram.py (hunk 7 above)
Comment (diehlbw) [reply src/seismometer/seismogram.py:73]: The consistency would be nice, but I'm not sure this is the best PR.  
Would you want a separate issue written up for tracking?  And does the change fit with any of the control-viz changes you are doing?

I'm reluctant to change it here, when I know we'll be focused more deeply on handling cohorts and cohort attributes soon.

File: src/seismometer/seismogram.py (hunk 8 above)
Comment (diehlbw) [reply src/seismometer/seismogram.py:288]: I started to do this but will defer with the same expectation as the cohort_cols.

This attribute doesn't seem to be accessed currently, indicating some more focused attention is probably warranted

File: docs/reference/seismogram.rst (hunk 9 above)
Comment (diehlbw) [reply docs/reference/seismogram.rst:21]: I was trying to set up my IDE to identify restructuredtext for basic autoformatting (primarily tabsize and trailing whitespace) and the main style guides I found said 3-space tabs for rst though it doesn't seem fully established.  
It seems like ours is _mostly_ 3, but some interspersed 4-space.

File: src/seismometer/seismogram.py (hunk 7 above)
Comment (gbowlin) [reply src/seismometer/seismogram.py:73]: Writing up as an separate issue for later resolution

File: src/seismometer/seismogram.py (hunk 8 above)
Comment (gbowlin) [reply src/seismometer/seismogram.py:288]: See #32

File: docs/reference/seismogram.rst (hunk 9 above)
Comment (gbowlin) [reply docs/reference/seismogram.rst:21]: Is there to formalize this as an expectation?

File: tests/test_startup.py [hunk 15]
```diff
@@ -10,20 +10,25 @@
...
+
 @pytest.fixture
 def fake_seismo(tmp_path):
-    with patch.object(Seismogram, "_load_from_config", fake_load):
-        Seismogram(config_path=tmp_path / "config", output_path=tmp_path / "output")
-    yield
+    with patch.object(Seismogram, "load_data", fake_load_data):
+        with patch.object(Seismogram, "load_config", fake_load_config):
+            Seismogram(config_path=tmp_path / "config", output_path=tmp_path / "output")
+        yield
```
Comment (gbowlin) [inline tests/test_startup.py:31]:
```suggestion

    with patch.object(Seismogram, "load_data", fake_load_data), \
         patch.object(Seismogram, "load_config", fake_load_config):
        Seismogram(config_path=tmp_path / "config", output_path=tmp_path / "output")
        yield
```

File: src/seismometer/data/loader/pipeline.py [hunk 16]
```diff
@@ -0,0 +1,158 @@
...
+        event_obj : pd.DataFrame, optional
+            an optional pre-loaded events dataframe, if None then will load from configuration, by default None.
+
+        Returns
+        -------
+        pd.DataFrame
+            the loaded and merged dataframe for a Seismogram session.
+        """
+
+        logger.info(f"Importing files from {self.config.config_dir}")
```
[the remark this answers is not in the document]
Comment (diehlbw) [reply src/seismometer/data/loader/pipeline.py:124]: Updated it to reveal configuration, not the action taken since that is no longer guaranteed

File: src/seismometer/data/loader/pipeline.py (hunk 1 above)
Comment (diehlbw) [reply src/seismometer/data/loader/pipeline.py:78]: Ammended here and event_fn

File: src/seismometer/data/loader/pipeline.py (hunk 2 above)
Comment (diehlbw) [reply src/seismometer/data/loader/pipeline.py:133]: Done

File: src/seismometer/data/loader/pipeline.py (hunk 3 above)
Comment (diehlbw) [reply src/seismometer/data/loader/pipeline.py]: Three of the four exposed objects are typing generics and not functions.  

The decorator would need fairly significant reworking to handle these, _or_ we could override both __name\_\_ and __module\_\_ similar to the docstring.  
That felt much worse than just doing the manual declaration.

**Review by gbowlin — DISMISSED** [review]
Left a few minor comments on your unit tests, but they are just documentation asks.

File: tests/data/loaders/test_load_events.py [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+
+# endregion
+# region File-type setup functions
+def parquet_setup():
+    file = Path("events.parquet")
+
+    data = event_frame()
+    data.to_parquet(file)
+
+    return fake_config(file)
```
[the remark this answers is not in the document]
Comment (diehlbw) [reply tests/data/loaders/test_load_events.py:58]: Added.

File: tests/test_startup.py (hunk 15 above)
Comment (diehlbw) [reply tests/test_startup.py:31]: black doesn't like the \ line continuation, but did change to be a single with

**Review by gbowlin — APPROVED** [review]
